modules/node-shared/src/main/scala/org/tessellation/node/shared/domain/cluster/programs/Joining.scala

```diff
@@ -190,6 +76,47 @@ sealed abstract class Joining[F[_]: Async: GenUUID: SecurityProvider: KryoSerial
     } yield ()
   }.onError(err => logger.error(err)(s"Error during join attempt by ${joinRequest.registrationRequest.id.show}"))
 
+  private def validateJoinConditions(): F[Unit] =
+    for {
+      nodeState <- nodeStorage.getNodeState
+      canJoinCluster <- nodeStorage.canJoinCluster
+      _ <- Applicative[F].unlessA(canJoinCluster)(NodeStateDoesNotAllowForJoining(nodeState).raiseError[F, Unit])
+    } yield ()
+
+  private def joinAndDiscover(context: P2PContext): F[Set[P2PContext]] = for {
+    _ <- twoWayHandshake(context, none).attempt.map(_.left.map { err =>
+      logger.error(err)(s"Joining to P2PContext ${context.show} failed")
+    })
```

Comment:
Per our initial discussion with @TheMMaciek , failure of the `twoWayHandshake` shouldn't stop the overall joining process. If `twoWayHandshake` fails, then `clusterStorage.getPeer(context.id)` returns a `None` and we move on.

Reply:
IMO if it's an initial joining it should fail, but for consecutive joins the fail should be ignored and discovery should continue

Reply:
Initial join being the node we send as a payload on the /join endpoint.

Reply:
In the same time I'm not sure if we need a notion of the initial node and joining to it and a separate join to the rest of the nodes, as not being able to join to the initial node will end up with 0 nodes joined and we can easily detect it as a post joining step and have some logic around it.

Reply:
If there are no joined `Peer`s, `PeerSelect` with throw an error, stopping the application. Is that insufficient? Is an additional check needed?

Reply:
I think we could check if we joined to at least a single initial node here (not sure if every layer uses `PeerSelect`). I would also think about aggregating `joined` and `failed to join` peers to have some sanity check before continuing to the next step. For the simplest initial implementation I think we need at least this first check (if the initial node join succeeded.)

Reply:
Node state trasition is now gated by a check of the `ClusterStorage`

```scala
...
    } >> clusterStorage.getResponsivePeers.flatMap { peers =>
      nodeStorage.tryModifyState(SessionStarted, stateAfterJoining).whenA(peers.nonEmpty)
    }
```

Reply:
@TheMMaciek , if the initial peer failed to join, then we cannot discover other peers from it, so the joining process ends (due to no new peers to attempt to join to) and node state transition fails because the `ClusterStorage` is empty.